addons/website_sale/controllers/main.py

```diff
@@ -1102,6 +1108,129 @@ def address(self, **kw):
         render_values.update(self._get_country_related_render_values(kw, render_values))
         return request.render("website_sale.address", render_values)
 
+    @http.route(
+        ['/shop/express'],
+        type='json',
+        methods=['POST'],
+        auth="public",
+        website=True,
+        sitemap=False,
+    )
+    def process_express_checkout(self, **post):
+        """ Records the partner information on the order when using express checkout flow.
+
+        Depending on whether the partner is registered and logged in, either creates a new partner
+        or uses an existing one that matches all received data.
+
+        :param dict post: a dictionary containing all the billing information sent by the express
+                          payment provider.
+        :return int: The order's partner id.
+        """
+        order = request.website.sale_get_order()
+
+        billing_address = post['billing_address']
+        # Update the partner with all the information
+        self._include_country_and_state_in_address(billing_address)
+
+        if order.partner_id.is_public:
+            self._convert_country_and_state_records_into_int(billing_address)
+            # If the partner doesn't exist yet, we create it.
+            billing_partner_id = self._create_or_edit_partner(
+                dict(billing_address, **{'type': 'invoice'})
+            )
+            order.partner_id = billing_partner_id
+            order.message_partner_ids = [Command.set([billing_partner_id])]
```

Comment:
It's a detail but this does the same thing:
```suggestion
            order.message_partner_ids = billing_partner_id
```